contribute/design/1.x/replica_scaleup/dataplane/20190910-replica-scaleup.md

```diff
@@ -0,0 +1,278 @@
+                  |  (istgtcontrol) 
+                 \|/  | 
+            CSTOR-VOLUME-MGMT <---->[CStorVolume CR]                          V
+                    | 
+                    | 
+                   \|/ 
+               [Conf file]
+```
+
+#### Sample YAMLs
+
+#### NOTES
+
+##### Is there a need to maintain the list of known replicas:
+This can be better explained with scenarios.
+1. Consider case where 3 replicas (R1, R2, R3) are connected and are fine.
+Lets say, one replica(R3) was lost and new one (R4) got added.
+After R4 got reconstructed with data, another replica (R2) got lost and
+new one (R5) got connected.
+Once R5 got reconstructed, all replicas have quorum propery on. But, just that
+R2 and R3 are not reachable.
+At a later point of time, if R1, R4, R5 are down and R2, R3 are connected to
+istgt (which can be due to attachment of old disks), istgt cannot identify that
+there is missing data with R2, R3.
+
+2. Consider replacing non-existing replica.
+When user reduced replication factor, and later increased it, if old replicas
+gets connected, istgt can't identify that there is missing data with replicas.
+
+### Low Level Design
+Current code takes care of reconstructing data to non-quorum replica once it
+does handshake with target. But, changes are required in allowing replica to
+perform handshake with target to achieve DesiredReplicationFactor.
+
+#### Replica Connection
+Below are the steps to allow replica handshake with istgt:
+- If currently connected replicas count >= 5, reject this if it is non-quorum or
+another connected non-quorum replica if the new one is in quorum
+- If DesiredRF number of replicas are healthy, reject
+- Make sure connected replica is an appropriate one to the volume, by checking
+replica name with vol name
+- If quorum is off, allow replica to connect and follow existing code flow
+(i.e, adding to non_quorum linked list etc)
+- If quorum is on, allow to connect only if it is new replica i.e.,
+its checkpointed nums are zero.
+- If quorum is on and its checkpointed nums are not zero, allow to connect only
+if it is in known lists, else reject
+
+
+#### Maintaining known replicas lists
+status.ReplicaList of CStorVolume CR need to contain the replicas information
+such that data can be correctly served from these replicas.
+There should be RF number of such replicas with quorum property 'on'.
+Such replicas are termed as 'Known Replicas'.
+
+For the case of newly created volume, all the RF number of replicas that connect
+to it with quorum ‘on’ can be added to known list. So,
+- Add the newly created replicas with quorum 'on' and checkpointed numbers as
+zero to known list by sending message to cstor-volume-mgmt.
+
+For the case of adding new replica or a replacment replica,
+- Identify the case when replica turned from quorum ‘off’ to quorum ‘on’ state.
+Let this replica referred as R.
+- Make sure there are no pending IOs on R [Why? If CR got updated, and there are
```

Comment:
Will this be blocking the IOs while the update to CR is happening?

Reply:
there will be pause of IOs. Implementation can be done in such a way that pause won't be more than specific time, lets say 5 seconds.